pystiche_papers/sanakoyeu_et_al_2018/_modules.py

```diff
@@ -184,3 +192,91 @@ def residual_block(channels: int) -> ResidualBlock:
     )
 
     return ResidualBlock(residual)
+
+
+def encoder(in_channels: int = 3,) -> SequentialEncoder:
+    r"""Encoder part of the :class:`Transformer` from :cite:`SKL+2018`.
+
+    Args:
+        in_channels: Number of channels in the input. Defaults to ``3``.
+
+    """
+    modules = [
+        nn.ReflectionPad2d(15),
+        ConvBlock(in_channels=in_channels, out_channels=32, kernel_size=3, stride=1),
+    ]
+    channels = (32, 32, 64, 128, 256)
+
+    modules.extend(
+        [
+            ConvBlock(
+                in_channels=in_channels,
+                out_channels=out_channels,
+                kernel_size=3,
+                stride=2,
+            )
+            for in_channels, out_channels in more_itertools.pairwise(channels)
+        ]
+    )
+
+    return SequentialEncoder(modules)
+
+
+def decoder(
+    out_channels: int = 3, num_residual_blocks: int = 9,
+) -> pystiche.SequentialModule:
+
+    modules: List[nn.Module] = []
+    for i in range(num_residual_blocks):
+        modules.append(residual_block(256))
+
+    channels = (256, 256, 128, 64, 32)
+    modules.extend(
+        [
+            UpsampleConvBlock(
+                in_channels=in_channels, out_channels=out_channels, kernel_size=3
+            )
+            for in_channels, out_channels in more_itertools.pairwise(channels)
+        ]
+    )
+    modules.append(nn.ReflectionPad2d(3))
+    modules.append(
+        conv(
+            in_channels=32,
+            out_channels=out_channels,
+            kernel_size=7,
+            stride=1,
+            padding="valid",
+        )
+    )
+    return pystiche.SequentialModule(*modules)
+
+
+class Decoder(nn.Module):
+    r"""Decoder part of the :class:`Transformer` from :cite:`SKL+2018`."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.decoder = decoder()
+        self.output_module = self.ValueRangeDelimiter()
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        return cast(torch.Tensor, self.output_module(self.decoder(input)))
+
+    class ValueRangeDelimiter(nn.Module):
```

Comment:
To take this even further: You can define this inside `__init__`. If you do, its no accessible afterwards, which is preferable here.